sw/host/rom_ext_image_tools/signer/src/main.rs

```diff
@@ -9,91 +9,177 @@
+        ),
+        (
+            &manifest::ROM_EXT_EXTENSION3_OFFSET,
+            &manifest::ROM_EXT_EXTENSION3_CHECKSUM,
+        ),
+    ];
+    for fields in &extensions {
+        image.set_manifest_field(fields.0, parse_hex_str("0x00000000")?)?;
+        image.set_manifest_field(fields.1, parse_hex_str("0x00000000")?)?;
+    }
+    image.set_manifest_field(
+        &manifest::ROM_EXT_USAGE_CONSTRAINTS,
+        std::array::IntoIter::new(
+            [0; manifest::ROM_EXT_USAGE_CONSTRAINTS.size_bytes],
+        ),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_PERIPHERAL_LOCKDOWN_INFO,
+        std::array::IntoIter::new(
+            [0; manifest::ROM_EXT_PERIPHERAL_LOCKDOWN_INFO.size_bytes],
+        ),
+    )?;
+    image.set_manifest_field(
+        &manifest::ROM_EXT_IMAGE_TIMESTAMP,
+        parse_hex_str("0x0000000000000000")?,
+    )?;
+
+    Ok(())
+}
 
-    // The whole image has been updated and signed, write the result to disk.
-    image.write_file();
+/// Calculates the signature for the signed portion of an image.
+fn calculate_image_signature(
+    image: &Image,
+    key: impl AsRef<Path>,
+) -> Result<ImageSignature> {
+    let key = fs::read(&key).with_context(|| {
+        format!("Failed to read key from `{}`.", key.as_ref().display())
+    })?;
+    let key =
+        PrivateKey::parse_from_der(&key).context("Failed to parse the key.")?;
+    let sig = ImageSignature::sign(&key, image.signed_bytes())
+        .context("Failed to sign the image.")?;
+    Ok(sig)
 }
 
-/// Generates the device usage value.
-///
-/// This value is extrapolated from the ROM_EXT manifest usage_constraints
-/// field, and does not reside in the ROM_EXT manifest directly.
-pub fn device_usage_value(path: &Path) -> Vec<u8> {
-    let _usage_constraints =
-        fs::read(path).expect("Failed to read usage constraints!");
-
-    // TODO: generate the device_usage_value from usage_constraints.
-    //       meanwhile use a "dummy" hard-coded vector.
-    vec![0xA5; 1024]
+/// Updates the signature of an image.
+fn update_image_signature(
+    image: &mut Image,
+    signature: ImageSignature,
+) -> Result<()> {
+    image.set_manifest_field(
+        &manifest::ROM_EXT_IMAGE_SIGNATURE,
+        signature.bytes().iter().rev().cloned(),
+    )?;
+    Ok(())
 }
 
-/// Obtains the system state value from the binary on disk.
-pub fn system_state_value(path: &Path) -> Vec<u8> {
-    fs::read(path).expect("Failed to read system state value!")
+fn main() -> Result<()> {
+    let args = Args::new(env::args_os())?;
+    let mut image = Image::from(fs::read(&args.input).with_context(|| {
+        format!("Failed to read the image from {}", args.input.display())
+    })?);
+
+    // TODO: This must use the public key.
+    update_image_manifest(&mut image, &args.priv_key)?;
+    let sig = calculate_image_signature(&image, &args.priv_key)?;
+    update_image_signature(&mut image, sig)?;
```

Comment:
TODO for a future refactor into `opentitantool`: These functions probably should belong to a `Signer` struct and should move into the `image` crate.

Reply:
Moved this note to `main.rs`.